gmso/formats/gsd.py

```diff
@@ -121,16 +121,18 @@ def _write_particle_information(
     unique_types = list(set(types))
     unique_types = sorted(unique_types)
     gsd_snapshot.particles.types = unique_types
-    warnings.warn("{} unique particle types detected".format(len(unique_types)))
+    warnings.warn(f"{len(unique_types)} unique particle types detected")
 
     typeids = np.array([unique_types.index(t) for t in types])
     gsd_snapshot.particles.typeid = typeids
 
     masses = np.array([site.mass for site in top.sites])
     masses[masses == 0] = 1.0
+    masses[masses == None] = 1.0
```

Comment:
Do you think we should either raise a warning here (that there is mass of 0/None which would be set to 1) or even raise and Error (that particle mass has not been initiated)?

Reply:
that's a good idea! I think we could have a method for Topology, just to scan through and make sure all necessary properties have been initiated, all objects have been parameterized, etc. This will not need to be explicitly called by the user.

Reply:
I guess we can have the warning/error here temporarily before the validating method is added to the main gmso.Topology

Reply:
Putting my two cents here, but I think we could do this with a custom unyt system per writer. Since masses are by default assumed to be in `gram/mol` in gmso. I don't like the conversion to numpy array here because we don't enforce any constraints on the units of a site mass.


```py

>>> from gmso import Atom, Topology
>>> import unyt as u
>>> import numpy as np
>>> top = Topology(name='mass demonstration')
>>> atom1 = Atom(mass=2.3 * u.gram / u.mol)
>>> atom2 = Atom(mass=2.4 * u.mg / u.mol)
>>> atom2.mass
unyt_quantity(2.4, 'mg/mol')
>>> atom1.mass
unyt_quantity(2.3, 'g/mol')
>>> top.add_site(atom1)
>>> top.add_site(atom2)
>>> top.update_topology()
>>> top
<Topology mass demonstration, 2 sites,
 0 connections,
 0 potentials,
 id: 140549244341696>
>>> mass_array = np.array([site.mass for site in top.sites])
>>> mass_array
array([2.3, 2.4])
```

Clearly, this is not correct. I think a common validation/filtration/customization  layer from between `gmso.core` and `gmso.formats` that can be leveraged by the writers can be developed (starting with the gsd writer). The idea is to define useful unit systems for hoomd/gsd at the writer level, however have these common utilities either in the validation/filteration layer or the gmso topology itself.

Reply:
Currently, I have been thinking of these writer/conversion, etc... as a series of dictionary transformations with differing levels of hierarchy per package/simulation engine.

Reply:
See https://github.com/mosdef-hub/gmso/issues/663 for a discussion on custom unyt systems

Reply:
These are good points @umesh-timalsina. Can I suggest that in this unyt validator, we can iterate through all properties with unyts attached to them and make sure they are equivalent? This would include the unyts in multiple different potential expressions attached to different typed objects across the topology. What we can do is raise an error if there are any differences, and then suggest a method available through the topology which converts all parameters following one of our available sets of base units. These same libraries than can be leveraged by particular writers for their necessary output styles.

Reply:
This is already the case with the potential expressions' expected dimensions in our potential template library. We had this idea of dropping unit object internally. However, this can be fleshed out further. This will take a substantial refactor of the existing codebase (internal part of the core API) however, what we need is a good set of default units to carry and some effort on the plumbing to make it seamless to the end user.

Reply:
>  The idea is to define useful unit systems for hoomd/gsd at the writer level

I think we should avoid defining writer specific unit systems. I think that adds unnecessary complications and potentially confusion for the end user.  For example, hoomd doesn't have a standard set of [units](https://hoomd-blue.readthedocs.io/en/v3.3.0/units.html). So in that case, if the user is using g/mol or mg/mol, or angstroms or nm in their topology, it doesn't matter. As long as the defined units are consistent; which I believe is what Cal is getting at.